src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -123,70 +132,80 @@ public boolean init(final Expression<?>[] vars, final int matchedPattern, final
 	
 	@Override
 	protected void execute(final Event e) {
-		if (!flattening) {
-			executeLegacy(e);
-			return;
-		}
-		
-		// 1.13 and newer: use BlockData
-		for (Block b : blocks.getArray(e)) {
-			BlockData data = b.getBlockData();
-			if (toggle == -1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(false);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(false);
-			} else if (toggle == 1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(true);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(true);
-			} else {
-				if (data instanceof Openable) // open = NOT was open
-					((Openable) data).setOpen(!((Openable) data).isOpen());
-				else if (data instanceof Powerable) // power = NOT power
-					((Powerable) data).setPowered(!((Powerable) data).isPowered());
+		ArrayList<Object> toggledValues = new ArrayList<>();
+		for (Object o : toggledExpr.getArray(e)) {
+			if (o instanceof Block) {
+				Block block = (Block) o;
+				if (!flattening) {
+					executeLegacy(block);
+					return;
+				}
+				
+				// 1.13 and newer: use BlockData
+				BlockData data = block.getBlockData();
+				if (toggle == -1) {
+					if (data instanceof Openable)
+						((Openable) data).setOpen(false);
+					else if (data instanceof Powerable)
+						((Powerable) data).setPowered(false);
+				} else if (toggle == 1) {
+					if (data instanceof Openable)
+						((Openable) data).setOpen(true);
+					else if (data instanceof Powerable)
+						((Powerable) data).setPowered(true);
+				} else {
+					if (data instanceof Openable) // open = NOT was open
+						((Openable) data).setOpen(!((Openable) data).isOpen());
+					else if (data instanceof Powerable) // power = NOT power
+						((Powerable) data).setPowered(!((Powerable) data).isPowered());
+				}
+				
+				block.setBlockData(data);
+				
+				toggledValues.add(block);
+				
+			} else if (o instanceof Boolean) {
+				toggledValues.add(!(Boolean) o);
 			}
-			
-			b.setBlockData(data);
 		}
+
+		toggledExpr.change(e, toggledValues.toArray(), ChangeMode.SET);
```

Comment:
You need to check that this expression allows changing, the parse-time check isn't enough in this case because it isn't always guaranteed to be checked (return type not boolean)

Reply:
But the check is already done via the ``init`` method, isn't it?

Reply:
It isn't always, it is only done if the return type is boolean (which is correct btw). You should add a check here verifying the expression can be changed. You should also only call the change method if there was a boolean in the input (otherwise changing isn't needed)

Reply:
``toggledValues`` are ``Block`` or ``Boolean``, so checking if it's a ``Boolean`` should not be a thing